warehouse/organizations/models.py

```diff
@@ -311,3 +341,140 @@ class OrganizationInvitation(db.Model):
 
     user = orm.relationship(User, lazy=False)
     organization = orm.relationship("Organization", lazy=False)
+
+
+class TeamRoleType(str, enum.Enum):
+
+    Member = "Member"
+
+
+class TeamRole(db.Model):
+
+    __tablename__ = "team_roles"
+    __table_args__ = (
+        Index("team_roles_user_id_idx", "user_id"),
+        Index("team_roles_team_id_idx", "team_id"),
+        UniqueConstraint(
+            "user_id",
+            "team_id",
+            name="_team_roles_user_team_uc",
+        ),
+    )
+
+    __repr__ = make_repr("role_name", "team", "user")
+
+    role_name = Column(
+        Enum(TeamRoleType, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    user_id = Column(
+        ForeignKey("users.id", onupdate="CASCADE", ondelete="CASCADE"), nullable=False
+    )
+    team_id = Column(
+        ForeignKey("teams.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+
+    user = orm.relationship(User, lazy=False)
+    team = orm.relationship("Team", lazy=False)
+
+
+class TeamProjectRoleType(str, enum.Enum):
+
+    Admin = "Admin"
+    Upload = "Upload"
+
+
+class TeamProjectRole(db.Model):
+
+    __tablename__ = "team_project_roles"
+    __table_args__ = (
+        Index("team_project_roles_project_id_idx", "project_id"),
+        Index("team_project_roles_team_id_idx", "team_id"),
+        UniqueConstraint(
+            "project_id",
+            "team_id",
+            name="_team_project_roles_project_team_uc",
+        ),
+    )
+
+    __repr__ = make_repr("role_name", "team", "project")
+
+    role_name = Column(
+        Enum(TeamProjectRoleType, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    project_id = Column(
+        ForeignKey("projects.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+    team_id = Column(
+        ForeignKey("teams.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+
+    project = orm.relationship("Project", lazy=False)
+    team = orm.relationship("Team", lazy=False)
+
+
+class TeamFactory:
+    def __init__(self, request, organization=None):
+        self.request = request
+        self.organization = organization
+
+    def __getitem__(self, name):
+        if self.organization is None:
+            organization = OrganizationFactory(self.request)[name]
+            return TeamFactory(self.request, organization)
+        try:
+            return (
+                self.request.db.query(Team)
+                .filter(
+                    Team.organization == self.organization,
+                    Team.normalized_name == func.normalize_pep426_name(name),
+                )
+                .one()
+            )
+        except NoResultFound:
+            raise KeyError from None
+
+
+# TODO: Do we enable HasEvents functionality? Or are these just Org Events?
```

Comment:
Since membership in teams can grant access, yes these should `HasEvents`

Reply:
Enabled `HasEvents` in commit [c7e2dcc](https://github.com/pypi/warehouse/pull/11665/commits/c7e2dcc84a6ea11e4947dba6fa929254b9037443).